add delta tests for reporting model update

lauramaxwell opened the pull request:
## Overview
<!--- What was done in the source branch -->
<!--- (i.e. bugfixes, feature additions, etc.) -->
- Add qualification tests to ensure reporting layer with new delta calculations are properly done (tests 15_1 and 15_2)
- Generalizes the calculation for the number of observations removed due to NAs in tests 9-13 so that changes to lSource don't impact qual tests

## Test Notes/Sample Code
<!--- Notes about testing or code to reproduce new functionality --->

## Connected Issues
<!--- Links to issues, using "Closes #NNN" if the issue is closed via PR --->

- Closes #24

jonthegeek reviewed (changes requested):
See my comments, plus https://github.com/Gilead-BioStats/gsm.qc/pull/29 for some tweaks to make things run prettier.

lauramaxwell commented on tests/testqualification/qualification/test-qual_T10_1.R:
```diff
@@ -7,7 +7,7 @@ outputs <- map(kri_workflows, ~ map_vec(.x$steps, ~ .x$output))
 testthat::test_that("Given appropriate raw participant-level data, a Data Change Rate Assessment can be done using the Normal Approximation method.", {
   # default ---------------------------------
-  test <- map(kri_workflows, ~ robust_runworkflow(.x, mapped_data))
+  test <- map(kri_workflows, ~ robust_runworkflow(.x, mapped_data) %>% suppressWarnings())
```
one issue with `expect_warning()` in this case is that not every KRI that is being mapped over produces the warning, so replacing `suppressWarnings()` with `expect_warning(., "value of 0 removed.")` results in an error. Do you know of a way to "accept" warnings but not necessarily "expect" them?

jonthegeek replied:
I think you can use `capture_warnings()` to grab warnings, filter out "value of 0 removed", and then see if there's anything left (and throw it as a warning if so). Or do something like `expect_length(filtered_warnings, 0)`.

lauramaxwell replied:
actually, if i put it outside of the `map` it works well

jonthegeek replied:
I get test failures when I try to run testqualification/test_qualification.R locally. If it's working how you need it it's probably ok, but I think something MIGHT be off.

jonthegeek reviewed (approved):
I'm approving because everything seems to work in GHA, but `source("tests/testqualification/test_qualification.R", echo = TRUE)` fails for me so let me know if you'd like me to figure out why!

lauramaxwell commented:
> > What failures are you getting? I don't get any failures locally- i'm wondering if it's package version discrepencies- or something else?
> 
> I installed dev versions of everything except gsm.core (because the others refer to @main for that so there are conflicts if I try to install @dev for that one), and I get a bunch of test failures, starting with this:
> 
> ```
> Failure (test-qual_T10_2.R:27:3): Data Change Rate Assessments can be done correctly using a grouping variable, such as Site, Country, or Study, when applicable.
> ... - removed (`actual`) not equal to nrow(test$kri0011$Analysis_Transformed) (`expected`).
> 
>   `actual`:    NA
> `expected`: 136.0
> ```
> 
> I got the same thing by installing @main of everything. I ran `pak::pak()` and then restarted my session to make sure everything is installed and up-to-date.
> 
> This is how I ran it this last time, to quiet it down a bit and make it easier to spot the warnings & errors (although I can see we still have a lot of messaging that doesn't use the logging system):
> 
> ```r
> testthat::set_max_fails(1)
> gsm.core::SetLogger(log4r::logger(
>   "WARN",
>   function(level, ...) {
>     msg <- log4r::bare_log_layout()(level, ...)
>     if (toupper(level) == "WARN") {
>       warning(msg)
>     } else {
>       stop(msg)
>     }
>   }
> ))
> source("tests/testqualification/test_qualification.R")
> ```

Huh. I'm not able to recreate your errors on my end. i added some logic to catch for situations where there are no warnings, in case that was causing an issue for you (just a guess based on the `NA` in the error you showed). I might see if others can try running locally to see how it goes.

jonthegeek commented:
@lauramaxwell There were also a bunch of warnings, including these two just before the failure:

```
Warning (test-qual_T10_2.R:8:3): Data Change Rate Assessments can be done correctly using a grouping variable, such as Site, Country, or Study, when applicable.
WARN7 values of [ GroupID ] with a [ Denominator ] value of 0 removed.

Backtrace:
     ▆
  1. └─purrr::map(...) at test-qual_T10_2.R:8:3
  2.   └─purrr:::map_("list", .x, .f, ..., .progress = .progress)
  3.     ├─purrr:::with_indexed_errors(...)
  4.     │ └─base::withCallingHandlers(...)
  5.     ├─purrr:::call_with_cleanup(...)
  6.     └─gsm.qc (local) .f(.x[[i]], ...)
  7.       └─global robust_runworkflow(.x, mapped_data, steps = 1:5)
  8.         └─(purrr::safely(~gsm.core::RunStep(lStep = steps, lData = lWorkflow$lData, ... at testqualification/qualification/qual_data.R:99:5
  9.           ├─purrr:::capture_error(.f(...), otherwise, quiet)
 10.           │ └─base::tryCatch(...)
 11.           │   └─base (local) tryCatchList(expr, classes, parentenv, handlers)
 12.           │     └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
 13.           │       └─base (local) doTryCatch(return(expr), name, parentenv, handler)
 14.           └─.f(...)
 15.             └─gsm.core::RunStep(lStep = steps, lData = lWorkflow$lData, lMeta = lWorkflow$meta)
 16.               ├─base::do.call(GetStrFunctionIfNamespaced(lStep$name), params)
 17.               └─gsm.core (local) `<fn>`(dfInput = `<df[,6]>`)
 18.                 └─gsm.core::LogMessage(...)
 19.                   └─log4r::warn(logger = .le$logger, level = toupper(level), message = message)
 20.                     └─global appender("WARN", ...)

Warning (test-qual_T10_2.R:10:3): Data Change Rate Assessments can be done correctly using a grouping variable, such as Site, Country, or Study, when applicable.
NAs introduced by coercion
Backtrace:
    ▆
 1. └─... %>% as.numeric() at test-qual_T10_2.R:10:3
 ```

lauramaxwell commented:
> @lauramaxwell There were also a bunch of warnings, including these two just before the failure:
> 
> ```
> Warning (test-qual_T10_2.R:8:3): Data Change Rate Assessments can be done correctly using a grouping variable, such as Site, Country, or Study, when applicable.
> WARN7 values of [ GroupID ] with a [ Denominator ] value of 0 removed.
> 
> Backtrace:
>      ▆
>   1. └─purrr::map(...) at test-qual_T10_2.R:8:3
>   2.   └─purrr:::map_("list", .x, .f, ..., .progress = .progress)
>   3.     ├─purrr:::with_indexed_errors(...)
>   4.     │ └─base::withCallingHandlers(...)
>   5.     ├─purrr:::call_with_cleanup(...)
>   6.     └─gsm.qc (local) .f(.x[[i]], ...)
>   7.       └─global robust_runworkflow(.x, mapped_data, steps = 1:5)
>   8.         └─(purrr::safely(~gsm.core::RunStep(lStep = steps, lData = lWorkflow$lData, ... at testqualification/qualification/qual_data.R:99:5
>   9.           ├─purrr:::capture_error(.f(...), otherwise, quiet)
>  10.           │ └─base::tryCatch(...)
>  11.           │   └─base (local) tryCatchList(expr, classes, parentenv, handlers)
>  12.           │     └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
>  13.           │       └─base (local) doTryCatch(return(expr), name, parentenv, handler)
>  14.           └─.f(...)
>  15.             └─gsm.core::RunStep(lStep = steps, lData = lWorkflow$lData, lMeta = lWorkflow$meta)
>  16.               ├─base::do.call(GetStrFunctionIfNamespaced(lStep$name), params)
>  17.               └─gsm.core (local) `<fn>`(dfInput = `<df[,6]>`)
>  18.                 └─gsm.core::LogMessage(...)
>  19.                   └─log4r::warn(logger = .le$logger, level = toupper(level), message = message)
>  20.                     └─global appender("WARN", ...)
> 
> Warning (test-qual_T10_2.R:10:3): Data Change Rate Assessments can be done correctly using a grouping variable, such as Site, Country, or Study, when applicable.
> NAs introduced by coercion
> Backtrace:
>     ▆
>  1. └─... %>% as.numeric() at test-qual_T10_2.R:10:3
> ```

OH! for me- warnings start with `\033[38;5;253m` and for you warnings start with `WARN`, so the gsub i do to pull out the number of NAs removed doesn't work for you. i can change the gsub to `gsub("\033\\[38;5;253m|WARN", "", .) to catch that. Or just grab to digits that end the first element of the warning.

jonthegeek commented:
@lauramaxwell Confirmed, that fixes it for me with all my bells and whistles turned back on!
